app/src/main/java/com/example/bottomnav/VariableContents.java

```diff
@@ -1,20 +1,121 @@
+            return makeOptFromDecl((CPPASTSimpleDeclSpecifier) dec.getDeclSpecifier(),
+                    (CPPASTDeclarator) dec.getDeclarators()[0]);
+        } else {
+            return makeOptFromDecl((CPPASTNamedTypeSpecifier) dec.getDeclSpecifier(),
+                    (CPPASTDeclarator) dec.getDeclarators()[0]);
+        }
+    }
+
+    /**
+     *
+     * @param simpleSpec
+     * @param declarator
+     * @return Optional<Components>
+     * @throws Exception
+     *
+     * Difference: Simple specifiers don't quantifier types and do not have name object
+     */
+    private static Optional<VariableContents> makeOptFromDecl(CPPASTSimpleDeclSpecifier simpleSpec,
+                                                              CPPASTDeclarator declarator) throws Exception {
+        CPPASTEqualsInitializer init = (CPPASTEqualsInitializer) declarator.getInitializer();
+        Optional<CPPASTEqualsInitializer> initOpt = makeOptInit(init);
+
+        if (declarator != null && simpleSpec != null) {
+            String declaratorNameStr = declarator.getName().getRawSignature();
+            String typeQualiferNameStr = simpleSpec.getSyntax().getImage();
+            String primitiveNameStr = getPrimitiveType(simpleSpec.getType());
+            return Optional.of(new VariableContents(declaratorNameStr, typeQualiferNameStr,
+                    primitiveNameStr, initOpt));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     *
+     * @param namedTypeSpec
+     * @param declarator
+     * @return Optional<Components>
+     * @throws Exception
+     *
+     * Difference: Name typed specifiers have quantifer types and have name object
+     */
+    private static Optional<VariableContents> makeOptFromDecl(CPPASTNamedTypeSpecifier namedTypeSpec,
+                                                              CPPASTDeclarator declarator) throws Exception {
+        CPPASTEqualsInitializer init = (CPPASTEqualsInitializer) declarator.getInitializer();
+        Optional<CPPASTEqualsInitializer> initOpt = makeOptInit(init);
+        if (declarator != null && namedTypeSpec != null) {
+            String declaratorNameStr = declarator.getName().getRawSignature();
+            String typeQualifierNameStr = namedTypeSpec.getSyntax().getImage();
+            String primitiveTypeNameStr = namedTypeSpec.getName().getRawSignature();
+            return Optional.of(new VariableContents(declaratorNameStr, typeQualifierNameStr,
+                    primitiveTypeNameStr, initOpt));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private static Optional<CPPASTEqualsInitializer> makeOptInit(CPPASTEqualsInitializer init) {
+        if (init != null) {
+            return Optional.of(init);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private Optional<String> getValue(Optional<CPPASTEqualsInitializer> givenInit) {
```

Comment:
This Optional to Optional feels a bit weird (non-orthogonality - the Optional wrapper should be handled by an upper layer).
Also, what is this used for?

Reply:
Consts have a value assigned to them, while struct variables don't. There are some cases where CPPASTEqualsInitializer doesn't exist, so I needed to check for that.

Reply:
I think what I mean here is that you should structure this function as CPPASTEqualsInitializer -> String, and deal with the Optional separately.

Take a look at `Optional.map`: https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

Essentially, instead of duplicating the function of `Optional.map` in this function, I'd recommend writing the function to do one thing, then compose it with `Optional.map` with the data you're working is wrapped in an Optional.